Change to native pgx/v5 driver with scany

lzap opened the pull request:
Replaces database/sql and sqlx with pgx driver and scany library. There are multiple reasons why I want to do this.

* The code is cleaner and more simple since pgx driver automatically prepares SQL statements and cache them.
* Better errors (error messages do appear to be more detailed).
* Native logging via zerolog and tracing via opentelemetry.
* The driver natively supports many custom types, including JSONB which we use (which should be faster to encode/decode).
* The driver of choice for image builder, this allows us to directly use jobqueue and we could drop our copy from the codebase.

This is a WIP branch, the way I want to approach this is to show how it looks like and go through review first, once the code is okay I will proceed with the all DAO facades. The only implemented facade is: Accounts (integration tests are passing for this DAO).

This uses very recent version of pgx (v5) and scanny (v2) and tern migration library (v2). While pgx/v5 has been out for few weeks, both mentioned libraries are currently in v2-beta stage. The `replace` statement in `go.mod` is temporary, I hope the pgx author will merge my small patch that enables to use proper Go mod: https://github.com/jackc/tern/pull/69

Done:

* [x] Simplify DAO errors
* [x] Remove NameForError functions from DAO
* [x] Remove Value/Scan from reservation model (comment `// TODO: Use pgx native driver with scany library instead of sqlx which does not`)

Things I want to add as separate commits after review is done:

* [x] Get rid of `error` from GetXXXDao methods

Things to be done in other PRs:

* [ ] Rename Pubkey Resource DAO methods to `UnscopedXXXResource`
* [ ] Refactor dejq to PGX
* [ ] Add Prometheus metrics (https://github.com/IBM/pgxpoolprometheus/pull/5)

lzap commented on cmd/pbapi/main.go:
```diff
@@ -54,7 +54,7 @@ func main() {
 
 	// initialize stdout logging and AWS clients first
-	log.Logger = logging.InitializeStdout()
+	logging.InitializeStdout()
```
The `log.Logger` is now set by the function itself.

lzap commented on cmd/pbapi/main.go:
```diff
@@ -76,10 +76,11 @@ func main() {
 	if err != nil {
 		log.Fatal().Err(err).Msg("Error initializing database")
 	}
+	defer db.Close()
```
Connection pool is now properly closed.

lzap commented on cmd/pbapi/main.go:
```diff
@@ -76,10 +76,11 @@ func main() {
 
 	// initialize the rest
-	err = db.Initialize("public")
+	err = db.Initialize(ctx, "public")
```
Database initialization and migration now requires `Context`.

lzap commented on go.mod:
```diff
@@ -2,6 +2,11 @@ module github.com/RHEnVision/provisioning-backend
+// go get github.com/jackc/tern/migrate@v2.0.0-beta.1
+
+replace github.com/jackc/tern => github.com/lzap/tern2 v0.0.0-20220813151204-54e00e2a8f9d
+
```
This will be removed once `tern` v2 is properly tagged, the two `go get` commands are required until stable versions are released tho.

I suggest to go ahead and merge this so we can start testing this on stage.

ezr-ondrej replied:
hump, we will need to convince our linter tho.
Also do we have any timelines for the releases.
I'm fine being early adopter, but only if we know the releases are likely to happen in timely manner.

lzap commented on internal/config/0_common_config.go:
```diff
@@ -26,6 +26,10 @@ func init() {
+	parser.Viper.SetDefault("database.minConn", 2)
+	parser.Viper.SetDefault("database.maxConn", 100)
+	parser.Viper.SetDefault("database.maxLifetime", 2*time.Hour)
+	parser.Viper.SetDefault("database.maxIdleTime", 15*time.Minute)
```
FYI the default values from the pgx driver are:

* Minimum conn: 0
* Max conn: 4
* Lifetime: 1 hour
* Idletime: 30 minutes

Our values are wild guesses, we need to talk with SRE about this in more detail.

lzap commented on internal/dao/tests/account_test.go:
```diff
@@ -108,7 +108,7 @@ func TestGetByOrgId(t *testing.T) {
+func TestGetOrCreateByIdentityGetAccount(t *testing.T) {
 	accDao, ctx := setupAccount(t)
 	defer teardownAccount(t)
 	account, err := accDao.GetOrCreateByIdentity(ctx, "1", "1")
```
Renamed these so I can test against "Account" pattern.

ezr-ondrej replied:
What is "Account" pattern? :thinking:

lzap replied:
You can run tests against a regexp patter, so I was running all tests starting with Account, then with Pubkey etc. This allowed me to refactor and test DAO after DAO.

lzap commented on internal/db/connection.go:
```diff
@@ -1,22 +1,30 @@
+var PGX *pgxpool.Pool
+
+// TODO remove after all DAO packages are rewritten
 var DB *sqlx.DB
```
This will be removed once done with the refactoring.

lzap commented on internal/db/connection.go:
```diff
@@ -37,41 +45,69 @@ func GetConnectionString(prefix, schema string) string {
+			poolConfig.ConnConfig.Tracer = &tracelog.TraceLog{
+				Logger:   zeroLogger,
+				LogLevel: logLevel,
+			}
```
Tracing and logging with full context support and spans, yay! :-)

lzap commented on internal/db/seeds/emed.go:
```diff
@@ -0,0 +1,6 @@
+import "embed"
+
+//go:embed *.sql
+var EmbeddedSeeds embed.FS
```
Had to move both migrations and seeds into the directory because the way the Tern API is designed today, it does not allow callers to locate files from subdirectories.

lzap commented:
Pushed new changes:

* Updated `go.mod` it no longer requires `replace`
* Pubkeys pgx implementation
* More pubkey tests - all are green now

Last thing to solve is reservations implementation, then I can remove sqlx completely. Tests will be failing until this is fixed.

I am thinking renaming `db.PGX` to `db.Pool` which is a better name and also PGX is kinda hard to type.

ezr-ondrej commented on internal/dao/tests/pubkey_resource_test.go:
```diff
@@ -74,6 +75,13 @@ func TestGetResourceByProviderType(t *testing.T) {
 	assert.Equal(t, resource, createdResource)
 }
 
+func TestGetPubkeyResourceByProviderTypeNoRows(t *testing.T) {
```
I'd rather see more test in one function by `t.Run()` if those are for the same function, I've tried to start this, but we do not have it everywhere, so up to you, but I'd love to have test function per function mapping 1:1 if possible.

lzap replied:
Sure.

ezr-ondrej commented on internal/dao/dao_errors.go:
```diff
@@ -8,8 +8,17 @@ import (
+// TODO let's vastly simplify error handling in DAO, let's just create aliases for native PGX errors
+// and that should be enough:
+
+var ErrNoRows = pgx.ErrNoRows
+var ErrAffectedMismatch = errors.New("unexpected affected rows")
```
I guess linter wants to group the var statements, right?

lzap replied:
Yeah.

ezr-ondrej reviewed (commented):
So if I understand correctly the pgx implementatios are tested, but not used in prod, right? :)
Just making sure, I agree it's a correct approach :+1:
I have just one nit and I'd love to hear your opinion on keeping the naming convention for test methods 1:1 with tested methods

lzap commented:
> So if I understand correctly the pgx implementatios are tested, but not used in prod, right? :)

Yeah, this is only during the review tho, at the end of this PR, I will remove sqlx completely. Having two implementations for even a week is a dangerous thing (we could accidentally change only one and not the other).

lzap commented:
The work is done, app is working properly locally.

This is now ready for formal review. After this is confirmed to be working, I want to add two more commits to further simplify and rename some DAO interface functions. These would be pure refactoring, no code changes.

lzap commented:
(The linter fails because it sees we use unreleased version of the tern library. Let's ignore that for a moment.)

lzap commented:
> Can these be done in separate PR then?

Sure, then consider this as done, feel free to review. I will disable the linters complain about use of unreleased versions, both project authors confirmed they intend to release soon.

lzap commented:
Rebased because tests were failing. I also renamed `PGX` to `Conn` which is better.

lzap commented on internal/dao/dao_interfaces.go:
```diff
@@ -71,6 +73,7 @@ type ReservationDao interface {
 
 	// ListInstances returns instances associated to a reservation. UNSCOPED.
+	// It currently lists all instances and not instances for a reservation, this is a TODO.
 	ListInstances(ctx context.Context, limit, offset int64) ([]*models.ReservationInstance, error)
```
https://issues.redhat.com/browse/HMSPROV-298

lzap commented on internal/dao/tests/pubkey_test.go:
```diff
@@ -44,90 +74,136 @@ func teardownPubkey(_ *testing.T) {
+		require.ErrorIs(t, err, dao.ErrAffectedMismatch)
+	})
+
+	// TODO validators
```
This TODO is for me to add even more pubkey tests (validators).

I can add it either during rebase of this PR or later.

ezr-ondrej reviewed (approved):
I've got two filosophical comments, but nothing major, I'm all for getting this in. Looks as great improvement :)

ezr-ondrej commented on internal/dao/pgx/account_pgx.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+func (x *accountDao) GetById(ctx context.Context, id int64) (*models.Account, error) {
+	query := `SELECT * FROM accounts WHERE id = $1 LIMIT 1`
```
I've gotten convinced these are better directly in the functions, as it's usually changed together.

lzap replied:
Yeah indeed, we can get to a point that some queries would be shared across functions, in that case let's totally make constants. But so far it's trivial.

lzap commented:
Merged DAO and DAO stubs errors into a single package, it's now consistent and clean. Added extra validation tests to the pubkey dao tests suite. Rebased.

ezr-ondrej reviewed (approved):
Great job @lzap ! let's hope the libraries will release soon, but our code looks much healthier now! :tada:
